metal/resource_metal_device.go

```diff
@@ -649,9 +697,59 @@ func resourceMetalDeviceUpdate(d *schema.ResourceData, meta interface{}) error {
 		}
 
 	}
+
+	if d.HasChange("operating_system") || d.HasChange("user_data") || d.HasChange("custom_data") {
+		reinstallOptions, err := getReinstallOptions(d)
+
+		if err != nil {
+			return friendlyError(err)
+		}
+
+		if _, err := client.Devices.Reinstall(d.Id(), &reinstallOptions); err != nil {
+			return friendlyError(err)
+		}
+
+		if err = waitForActiveDevice(d, meta); err != nil {
+			return err
+		}
+	}
+
 	return resourceMetalDeviceRead(d, meta)
 }
 
+func getReinstallOptions(d *schema.ResourceData) (packngo.DeviceReinstallFields, error) {
+	reinstall_list, ok := d.Get("reinstall").([]interface{})
+
+	if !ok {
+		return packngo.DeviceReinstallFields{}, fmt.Errorf("expected reinstall configuration and none available")
+	}
+
+	if len(reinstall_list) == 0 {
+		return packngo.DeviceReinstallFields{}, fmt.Errorf("expected reinstall configuration and none available")
+	}
+
+	reinstall_config, ok := reinstall_list[0].(map[string]interface{})
+
+	if !ok {
+		return packngo.DeviceReinstallFields{}, fmt.Errorf("expected reinstall configuration and none available")
+	}
+
+	preserveData, ok := reinstall_config["preserve_data"].(bool)
+	if !ok {
```

Comment:
The `!ok` blocks are redundant, the value here and for deprovisionFast will already be false.
We can `_` discard `ok`.